src/deluge/processing/vector_rendering_function.h

```diff
@@ -15,92 +15,122 @@
  * If not, see <https://www.gnu.org/licenses/>.
  */
 
+#include <arm_neon.h>
+#include <cstddef>
+#include <cstdint>
+
 // Hard-coded "for-loop" for the below function.
-#define waveRenderingFunctionGeneralForLoop(i)                                                                         \
-	{                                                                                                                  \
-		phaseTemp += phaseIncrement;                                                                                   \
-		uint32_t rshifted = phaseTemp >> (32 - 16 - tableSizeMagnitude);                                               \
-		strength2 = vset_lane_u16(rshifted, strength2, i);                                                             \
-                                                                                                                       \
-		uint32_t whichValue = phaseTemp >> (32 - tableSizeMagnitude);                                                  \
-		uint32_t* readAddress = (uint32_t*)((uint32_t)table + (whichValue << 1));                                      \
-                                                                                                                       \
-		readValue = vld1q_lane_u32(readAddress, readValue, i);                                                         \
-	}
+template <int i>
+[[gnu::always_inline]] static inline void
+waveRenderingFunctionGeneralForLoop(uint32x4_t& readValue, uint16x4_t& strength2, uint32_t& phaseTemp,
```

Comment:
I think that's kinda orthagonal to the problem, since I could have just added another parameter to take the actual function (or wrapped it in a function templated over the input function types that returns a lambda that calls the given function). The reason the macro is still a lambda instead of simply placed in-source because it's used in two places, but those two places have different contextual references that need to be captured by `[&]`

Yeah, actually, I'll just inline this and include a reference to what it originally was.

Reply:
Also, from what I know, for C++11 lambdas were originally built off of syntax sugar for templates, so they should be taking the same paths in the optimizer